### PR TITLE
end-to-end tests for resource groups

### PR DESCRIPTION
**Pull request by vusirikala**
### Description

End to end tests to test resource groups. The tests run a bunch of operations on resource groups with RESOURCE_GROUPS_CHARGE_AS_SIZE_SUM enabled and disabled and compare that the results are equal.

### CODE REVIEW

File: aptos-move/e2e-move-tests/src/tests/resource_groups.data/pack/sources/resource_groups_test.move [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+    const ERESOURCE_GROUP_DOESNT_EXIST: u64 = 21;
+    const EINDEX_DOESNT_EXIST: u64 = 22;
+    const EOPTION_DOESNT_EXIST: u64 = 23;
+
+    #[resource_group(scope = global)]
+    struct MyGroup {}
+
+    // TODO: I choose all these resources to have different structure to have a more comprehensive testing.
+    // But this also means, I needed to have individual set functions for each resource. Is that okay?
+    // Or should I have same structure for all the resources?
```
Comment (igor-aptos) [inline aptos-move/e2e-move-tests/src/tests/resource_groups.data/pack/sources/resource_groups_test.move:28]: I think it is perfectly fine to have all resources have a string field (potentially at different indices), which is set via the function, and have the rest be different, and just be initialized.

we are generally testing behavior across resources, it is unlikely a single resource itself might have issues.


and then you don't need different set methods.


you also probably need a check method - i.e. comparison to a value, and aborting if not matching, and then potentially some methods that touch two things like set_and_check(index1, index2)

File: aptos-move/e2e-move-tests/proptest-regressions/tests/resource_groups.txt [hunk 2]
```diff
@@ -0,0 +1,7 @@
+# Seeds for failure cases proptest has generated in the past. It is
```
[the remark this answers is not in the document]
Comment (vusirikala) [reply aptos-move/e2e-move-tests/proptest-regressions/tests/resource_groups.txt]: Done.

File: aptos-move/e2e-move-tests/src/tests/resource_groups.data/pack/sources/resource_groups_test.move (hunk 1 above)
Comment (vusirikala) [reply aptos-move/e2e-move-tests/src/tests/resource_groups.data/pack/sources/resource_groups_test.move]: Done.

**Review by georgemitenkov — APPROVED** [review]

File: aptos-move/e2e-move-tests/src/tests/resource_groups.data/pack/sources/resource_groups_test.move [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+
+    public entry fun init_signer(main_account: &signer, seed: vector<u8>) {
+        let (_resource_account_signer, signer_cap) = account::create_resource_account(main_account, seed);
+        let main_resource = MainResource {
+            signer_cap,
+        };
+        move_to<MainResource>(main_account, main_resource);
+    }
+
+    public entry fun set_resource(_delegated_signer: &signer, main_account: address, index: u32, name: String, value: u32) acquires MainResource, MyResource1, MyResource2, MyResource3, MyResource4 {
```
[the remark this answers is not in the document]
Comment (vusirikala) [reply aptos-move/e2e-move-tests/src/tests/resource_groups.data/pack/sources/resource_groups_test.move:49]: Done.

File: aptos-move/e2e-move-tests/src/tests/aggregator_v2.rs [hunk 4]
```diff
@@ -117,7 +117,7 @@ mod test_cases {
 }
 
 #[allow(dead_code)]
```
[the remark this answers is not in the document]
Comment (vusirikala) [reply aptos-move/e2e-move-tests/src/tests/aggregator_v2.rs:119]: Removed it.

File: aptos-move/e2e-move-tests/src/resource_groups.rs [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+    }
+
+    pub fn set_and_check(&mut self, index1: u32, index2: u32, name1: String, value1: u32, name2: String, value2: u32) -> SignedTransaction {
+        self.txn_index += 1;
+        self.harness.create_entry_function(
+            &self.txn_accounts[self.txn_index % self.txn_accounts.len()],
+            str::parse("0x1::resource_groups_test::set_and_check").unwrap(),
+            vec![],
+            vec![
+                bcs::to_bytes(self.account.address()).unwrap(),
```
[the remark this answers is not in the document]
Comment (vusirikala) [reply aptos-move/e2e-move-tests/src/resource_groups.rs:220]: Yes. The signer is the one who is submitting the transaction. Who choose different signers for each transaction. `bcs::to_bytes(self.account.address()).unwrap()` is the fixed address that is storing the resource group. This address is fixed for all the transactions.

File: aptos-move/e2e-move-tests/src/tests/resource_groups.rs [hunk 6]
```diff
@@ -1,13 +1,153 @@
...
+};
+use aptos_language_e2e_tests::executor::ExecutorMode;
 use aptos_package_builder::PackageBuilder;
 use aptos_types::{account_address::AccountAddress, on_chain_config::FeatureFlag};
 use move_core_types::{identifier::Identifier, language_storage::StructTag, vm_status::StatusCode};
+use proptest::prelude::*;
 use serde::Deserialize;
 use test_case::test_case;
 
+// This mode describes whether to enable or disable RESOURCE_GROUPS_CHARGE_AS_SIZE_SUM flag
```
[the remark this answers is not in the document]
Comment (vusirikala) [reply aptos-move/e2e-move-tests/src/tests/resource_groups.rs:20]: Added a TODO

File: aptos-move/e2e-move-tests/src/tests/resource_groups.rs [hunk 7]
```diff
@@ -1,13 +1,153 @@
...
+    })]
+
+    #[test]
+    fn proptest_resource_groups_1(test_env in arb_test_env(17)) {
+        println!("Testing test_aggregator_lifetime {:?}", test_env);
+        let mut h = setup(test_env.executor_mode, test_env.resource_group_mode, 17);
+
+        let txns = vec![
+            (SUCCESS, h.init_signer(vec![5,2,3])),
+            (SUCCESS, h.set_resource(4, "ABC".to_string(), 10)),
```
[the remark this answers is not in the document]
Comment (vusirikala) [reply aptos-move/e2e-move-tests/src/tests/resource_groups.rs:102]: Done.

File: aptos-move/e2e-move-tests/src/tests/resource_groups.data/pack/Move.toml [hunk 8]
```diff
@@ -0,0 +1,6 @@
+[package]
+name = "resource_groups_test"
+version = "0.0.0"
+
+[dependencies]
+AptosFramework = { local = "../../../../../framework/aptos-framework" }
```
[the remark this answers is not in the document]
Comment (vusirikala) [reply aptos-move/e2e-move-tests/src/tests/resource_groups.data/pack/Move.toml]: Done.

**Pull request merged**